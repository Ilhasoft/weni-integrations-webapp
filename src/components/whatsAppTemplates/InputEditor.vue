<template>
  <div class="input-editor">
    <div v-click-outside-element="hideEmoji">
      <VEmojiPicker
        v-show="displayEmoji"
        class="input-editor__emoji-picker"
        :emojiSize="22"
        :emojisByRow="7"
        @select="selectEmoji"
      />
      <unnnic-button
        class="input-editor__emoji-picker__button"
        type="terciary"
        iconCenter="emoji"
        size="small"
        @click="toggleEmoji"
      />
    </div>
    <div v-if="formatter" class="input-editor__formatter">
      <unnnic-tool-tip side="top" text="Bold" enabled>
        <unnnic-button
          class="input-editor__button"
          type="terciary"
          iconCenter="text-bold"
          size="small"
          @click="$emit('format-event', '*')"
        />
      </unnnic-tool-tip>

      <unnnic-tool-tip side="top" text="Italic" enabled>
        <unnnic-button
          class="input-editor__button"
          type="terciary"
          iconCenter="text-italic"
          size="small"
          @click="$emit('format-event', '_')"
        />
      </unnnic-tool-tip>

      <unnnic-tool-tip side="top" text="Strike Through" enabled>
        <unnnic-button
          class="input-editor__button input-editor__button__strike"
          type="terciary"
          size="small"
          @click="$emit('format-event', '~')"
        >
          <img :src="strikeIcon" />
        </unnnic-button>
      </unnnic-tool-tip>

      <unnnic-tool-tip side="top" text="Monospace" enabled>
        <unnnic-button
          class="input-editor__button input-editor__button__mono"
          type="terciary"
          size="small"
          @click="$emit('format-event', '```')"
        >
          {{ monospaceString }}
        </unnnic-button>
      </unnnic-tool-tip>
    </div>
    <unnnic-button
      type="terciary"
      iconLeft="add-1"
      size="small"
      class="templates-form-body__add-variable"
      @click="$emit('add-variable')"
    >
      {{ $t('WhatsApp.templates.form_field.add_variable') }}
    </unnnic-button>
  </div>
</template>

<script>
  import { VEmojiPicker } from 'v-emoji-picker';
  import StrikeThroughIcon from '@/assets/svgs/strike-through.svg';

  export default {
    name: 'InputEditor',
    components: {
      VEmojiPicker,
    },
    data() {
      return {
        displayEmoji: false,
        search: '',
        strikeIcon: StrikeThroughIcon,
        monospaceString: '</>',
      };
    },
    props: {
      formatter: {
        type: Boolean,
        default: true,
      },
    },
    methods: {
      hideEmoji() {
        this.displayEmoji = false;
      },
      toggleEmoji() {
        this.displayEmoji = !this.displayEmoji;
      },
      selectEmoji(emoji) {
        this.$emit('emoji-event', emoji);
        this.displayEmoji = false;
      },
    },
  };
</script>

<style lang="scss" scoped>
  .input-editor {
    display: flex;
    align-items: center;
    justify-content: flex-end;
    position: relative;

    &__formatter {
      display: flex;
    }

    &__button {
      width: $unnnic-icon-size-lg;
      height: $unnnic-icon-size-lg;

      &__strike {
        ::v-deep {
          span {
            align-self: baseline;
            opacity: 0.5;
          }
        }
      }

      &__mono {
        font-family: $unnnic-font-family-primary;
        font-size: $unnnic-font-size-body-lg;
        color: $unnnic-color-neutral-cleanest;
      }

      ::v-deep {
        svg {
          width: $unnnic-icon-size-sm;
          margin-left: $unnnic-spacing-inline-nano;
        }
      }
    }

    &__emoji-picker {
      position: absolute;
      left: 0px;
      z-index: 1;
      top: -270px;

      &__button {
        opacity: $unnnic-opacity-level-clarifying;
      }

      ::v-deep .container-emoji {
        height: 200px;
      }
    }
  }
</style>
