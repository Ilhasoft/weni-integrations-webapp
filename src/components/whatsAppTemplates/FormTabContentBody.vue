<template>
  <div class="form-tab-content-body">
    <span class="form-tab-content-body__title">
      {{ $t('WhatsApp.templates.form_field.body_text') }}
    </span>

    <unnnic-text-area
      ref="bodyText"
      :key="bodyKey"
      class="form-tab-content-body__input"
      :disabled="disableInputs"
      :value="bodyContent"
      :placeholder="$t('WhatsApp.templates.form_field.body_text__placeholder')"
      @input="onInput"
      :maxLength="1024"
      :type="hasErrors ? 'error' : 'normal'"
      :errors="errorsList"
    />
    <InputEditor
      :class="[
        'form-tab-content-body__input__actions',
        hasErrors && 'form-tab-content-body__input__actions--error',
      ]"
      :formatter="true"
      @format-event="handleFormatEvent"
      @add-variable="addVariable"
      @emoji-event="handleNewEmoji"
    />
    <unnnic-button type="tertiary" iconLeft="add-1" size="small" @click="addVariable">
      {{ $t('WhatsApp.templates.form_field.add_variable') }}
    </unnnic-button>
  </div>
</template>

<script>
  import InputEditor from '@/components/whatsAppTemplates/InputEditor';

  import {
    countVariables,
    startsWithVariableRegex,
    endsWithVariableRegex,
    singleBracketVariableRegex,
    incompleteStartBracketVariableRegex,
    incompleteEndBracketVariableRegex,
  } from '@/utils/countTemplateVariables.js';

  export default {
    name: 'FormTabContentBody',
    components: {
      InputEditor,
    },
    props: {
      disableInputs: {
        type: Boolean,
        default: false,
      },
    },
    data() {
      return {
        bodyKey: 0,
      };
    },
    computed: {
      templateTranslationCurrentForm(){
        return whatsapp_store().templateTranslationCurrentForm
      },
      bodyContent() {
        return this.templateTranslationCurrentForm?.body || '';
      },
      hasErrors() {
        return this.errorsList.length > 0;
      },
      errorsList() {
        const errors = [];

        if (
          this.bodyContent.match(startsWithVariableRegex) ||
          this.bodyContent.match(endsWithVariableRegex)
        ) {
          errors.push(this.$t('WhatsApp.templates.error.start_or_end_with_variable'));
        }

        if (
          this.bodyContent.match(singleBracketVariableRegex) ||
          this.bodyContent.match(incompleteStartBracketVariableRegex) ||
          this.bodyContent.match(incompleteEndBracketVariableRegex)
        ) {
          errors.push(this.$t('WhatsApp.templates.error.incomplete_bracket_variable'));
        }

        const variableCount = countVariables(this.bodyContent);
        const wordCount = this.countWords(this.bodyContent);
        if (wordCount && variableCount * 2 + 1 > wordCount - variableCount) {
          errors.push(this.$t('WhatsApp.templates.error.too_many_variables'));
        }

        return errors;
      },
    },
    methods: {
      /* istanbul ignore next */
      handleFormatEvent(eventCharacter) {
        if (this.disableInputs) {
          return;
        }

        const textArea = Array.from(this.$refs.bodyText.$el.children).find(
          (child) => child.nodeName === 'TEXTAREA',
        );

        const before = textArea.value.substring(0, textArea.selectionStart);
        const selectionContent = textArea.value.substring(
          textArea.selectionStart,
          textArea.selectionEnd,
        );
        const after = textArea.value.substring(textArea.selectionEnd);
        const result = `${before}${eventCharacter}${selectionContent}${eventCharacter}${after}`;

        this.emitInputChange(result);

        textArea.focus();
      },
      addVariable() {
        if (this.disableInputs) {
          return;
        }

        const variableCount = countVariables(this.templateTranslationCurrentForm?.body);

        let body = '';
        if (this.templateTranslationCurrentForm?.body) {
          body = this.templateTranslationCurrentForm.body.trim();
        }
        const fieldValue = (body + ` {{${variableCount + 1}}}`).trim();

        if (fieldValue.length >= 1024) {
          return;
        }

        this.emitInputChange(fieldValue);

        const textArea = Array.from(this.$refs.bodyText.$el.children).find(
          (child) => child.nodeName === 'TEXTAREA',
        );
        textArea.focus();
      },
      /* istanbul ignore next */
      onInput(event) {
        if (this.disableInputs) {
          return;
        }

        this.emitInputChange(event);
      },
      handleNewEmoji(emoji) {
        if (this.disableInputs) {
          return;
        }
        const result = (this.templateTranslationCurrentForm.body || '') + emoji.data;

        this.emitInputChange(result);
      },
      emitInputChange(fieldValue) {
        this.$emit('input-change', {
          fieldName: 'body',
          fieldValue,
        });

        this.$emit('input-change', {
          fieldName: 'bodyHasError',
          fieldValue: this.hasErrors,
        });
      },
      countWords(text) {
        if (!text) {
          return 0;
        }

        return text.trim().split(/\s+/).length;
      },
    },
  };
</script>

<style lang="scss" scoped>
  .form-tab-content-body {
    display: flex;
    flex-direction: column;

    &__title {
      margin-bottom: $unnnic-spacing-stack-sm;
      font-size: $unnnic-font-size-body-lg;
      line-height: $unnnic-font-size-body-lg + $unnnic-line-height-md;
      font-weight: $unnnic-font-weight-bold;

      color: $unnnic-color-neutral-darkest;
    }

    &__input {
      ::v-deep textarea {
        resize: none;
      }

      &__actions {
        margin-left: auto;
        margin-top: -28px;
        margin-right: $unnnic-spacing-inline-giant;

        &--error {
          margin-top: 0;
          margin-right: 0;
        }
      }
    }
  }
</style>import { whatsapp_store } from '@/stores/modules/appType/channels/whatsapp.store';

