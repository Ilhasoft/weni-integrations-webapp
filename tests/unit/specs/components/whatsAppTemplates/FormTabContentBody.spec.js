import Vuex from 'vuex';
import { mount, createLocalVue } from '@vue/test-utils';
import FormTabContentBody from '@/components/whatsAppTemplates/FormTabContentBody.vue';
import InputEditor from '@/components/whatsAppTemplates/InputEditor.vue';
import '@weni/unnnic-system';

const localVue = createLocalVue();
localVue.use(Vuex);

const mountComponent = ({ body = '', disableInputs = false } = {}) => {
  const getters = {
    templateTranslationCurrentForm: jest.fn(() => {
      return { body };
    }),
  };

  const store = new Vuex.Store({
    modules: {
      WhatsApp: {
        namespaced: true,
        getters,
      },
    },
  });

  const wrapper = mount(FormTabContentBody, {
    localVue,
    store,
    mocks: {
      $t: () => 'some specific text',
    },
    stubs: {
      InputEditor: true,
    },
    propsData: {
      disableInputs,
    },
  });

  return { wrapper, getters };
};

describe('components/whatsAppTemplates/FormTabContentBody.vue', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should be rendered properly', () => {
    const { wrapper } = mountComponent();
    expect(wrapper).toMatchSnapshot();
  });

  it('should call format event on inputEditor event', async () => {
    const spy = spyOn(FormTabContentBody.methods, 'handleFormatEvent');
    const { wrapper } = mountComponent({ body: 'Text {{1}} with {{2}} variables' });
    const inputEditor = wrapper.findComponent(InputEditor);

    inputEditor.vm.$emit('format-event', 'bold');
    await wrapper.vm.$nextTick();

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith('bold');
  });

  it('should not add variable if disabledInputs is true', async () => {
    const lengthyText =
      'Nam quis nulla. Integer malesuada. In in enim a arcu imperdiet malesuada. Sed vel lectus. Donec odio urna, tempus molestie, porttitor ut, iaculis quis, sem. Phasellus rhoncus. Aenean id metus id velit ullamcorper pulvinar. Vestibulum fermentum tortor id mi. Pellentesque ipsum. Nulla non arcu lacinia neque faucibus fringilla. Nulla non lectus sed nisl molestie malesuada. Proin in tellus sit amet nibh dignissim sagittis. Vivamus luctus egestas leo. Maecenas sollicitudin. Nullam rhoncus aliquam metus. Etiam egestas wisi a erat. Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Nullam feugiat, turpis at pulvinar vulputate, erat libero tristique tellus, nec bibendum odio risus sit amet ante. Aliquam erat volutpat. Nunc auctor. Mauris pretium quam et urna. Fusce nibh. Duis risus. Curabitur sagittis hendrerit ante. Aliquam erat volutpat. Vestibulum erat nulla, ullamcorper nec, rutrum non, nonummy ac, erat. Duis condimentum augue id magna semper rutrum. Nullam justo enim, consectetuer nec, ullamcorper ac, v';
    const { wrapper } = mountComponent({
      body: lengthyText,
    });

    const inputEditor = wrapper.findComponent(InputEditor);
    const input = wrapper.getComponent({ ref: 'bodyText' });
    expect(input.text()).toEqual(lengthyText);

    inputEditor.vm.$emit('add-variable');
    await wrapper.vm.$nextTick();

    expect(input.text()).toEqual(lengthyText);
    const event = wrapper.emitted('input-change');
    expect(event).toBeFalsy();
  });

  it('should not add variable if body will be greater than 1024', async () => {
    const { wrapper } = mountComponent({ body: 'text', disableInputs: true });

    const inputEditor = wrapper.findComponent(InputEditor);
    const input = wrapper.getComponent({ ref: 'bodyText' });
    expect(input.text()).toEqual('text');

    inputEditor.vm.$emit('add-variable');
    await wrapper.vm.$nextTick();

    expect(input.text()).toEqual('text');
    const event = wrapper.emitted('input-change');
    expect(event).toBeFalsy();
  });

  it('should add variable on inputEditor event', async () => {
    const { wrapper } = mountComponent({ body: 'Text {{1}} with {{2}} variables' });
    const inputEditor = wrapper.findComponent(InputEditor);
    const input = wrapper.getComponent({ ref: 'bodyText' });

    expect(input.text()).toEqual('Text {{1}} with {{2}} variables');

    inputEditor.vm.$emit('add-variable');
    await wrapper.vm.$nextTick();

    expect(input.text()).toEqual('Text {{1}} with {{2}} variables {{3}}');

    const event = wrapper.emitted('input-change');
    expect(event).toBeTruthy();
    expect(event.length).toBe(1);
    expect(event[0]).toEqual([
      {
        fieldName: 'body',
        fieldValue: 'Text {{1}} with {{2}} variables {{3}}',
      },
    ]);
  });

  it('should add variable on inputEditor event even without body', async () => {
    const { wrapper } = mountComponent({ body: undefined });
    const inputEditor = wrapper.findComponent(InputEditor);
    const input = wrapper.getComponent({ ref: 'bodyText' });

    expect(input.text()).toEqual('');

    inputEditor.vm.$emit('add-variable');
    await wrapper.vm.$nextTick();

    expect(input.text()).toEqual('{{1}}');

    const event = wrapper.emitted('input-change');
    expect(event).toBeTruthy();
    expect(event.length).toBe(1);
    expect(event[0]).toEqual([
      {
        fieldName: 'body',
        fieldValue: '{{1}}',
      },
    ]);
  });
});
