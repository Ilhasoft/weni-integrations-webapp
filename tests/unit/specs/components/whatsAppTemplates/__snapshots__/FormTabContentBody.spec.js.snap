// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/whatsAppTemplates/FormTabContentBody.vue should be rendered properly 1`] = `
<div class="form-tab-content-body"><span class="form-tab-content-body__title">
    Body Text
  </span>
  <div data-v-2f3d79bf="" class="form-tab-content-body__input unnnic-text-area md normal">
    <!----><textarea data-v-2f3d79bf="" placeholder="Write the body text" maxlength="1024"></textarea>
    <div data-v-2f3d79bf="" class="helper"> 0/1024 </div>
    <!---->
  </div>
  <inputeditor-stub formatter="true" class="form-tab-content-body__input__actions"></inputeditor-stub> <button data-v-2d9f2cbf="" class="unnnic-button unnnic-button--size-small unnnic-button--tertiary">
    <!----><span data-v-ab84ef7c="" data-v-2d9f2cbf="" class="unnnic-icon unnnic-icon__size--sm unnnic-icon-scheme--neutral-dark unnnic-button__icon-left"><svg viewBox="0 0 40 40" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M20 35.0005C19.4825 35.0005 19.0625 34.5805 19.0625 34.063V20.938H5.9375C5.42 20.938 5 20.518 5 20.0005C5 19.483 5.42 19.063 5.9375 19.063H19.0625V5.93799C19.0625 5.42049 19.4825 5.00049 20 5.00049C20.5175 5.00049 20.9375 5.42049 20.9375 5.93799V19.063H34.0625C34.58 19.063 35 19.483 35 20.0005C35 20.518 34.58 20.938 34.0625 20.938H20.9375V34.063C20.9375 34.5805 20.5175 35.0005 20 35.0005Z" fill="#3B414D" class="primary"></path></svg></span>
    <!----><span data-v-2d9f2cbf="" class="unnnic-button__label">
    Add Variable to text body
    </span>
    <!----></button>
</div>
`;
